File: medium/python/151.py
# https://leetcode.com/problems/reverse-words-in-a-string/?envType=study-plan-v2&envId=leetcode-75
# Time and Space: O(n)
class Solution:
    def reverseWords(self, s: str) -> str:
        wordsArray = s.split(" ")

        firstWordIndex = None
        lastWordIndex = None

        for index in range(len(wordsArray)):
            if wordsArray[index] != "":
                firstWordIndex = index
                break

        last = len(wordsArray) - 1
        while last >= 0:
            if wordsArray[last] != "":
                lastWordIndex = last
                break
            last = last - 1

        answerArray = []
        while last >= firstWordIndex:
            if wordsArray[last] != "":
                answerArray.append(wordsArray[last])
                if last != firstWordIndex:
                    answerArray.append(" ")
            last = last - 1

        return "".join(answerArray)

File: medium/ts/151.ts
// https://leetcode.com/problems/reverse-words-in-a-string/?envType=study-plan-v2&envId=leetcode-75

// Time: O(n)
// Space: O(n)
function reverseWords(s: string): string {
  let firstWord;
  let lastWord;
  let wordsArray = s.split(" ");
  for (let i = 0; i < wordsArray.length; i++) {
    if (wordsArray[i] !== "") {
      firstWord = i;
      break;
    }
  }

  for (let i = wordsArray.length - 1; i >= 0; i--) {
    if (wordsArray[i] !== "") {
      lastWord = i;
      break;
    }
  }

  let answerArray: string[] = [];
  let j = lastWord;
  while (j >= firstWord) {
    if (wordsArray[j] !== "") {
      answerArray.push(wordsArray[j]);
      if (j !== firstWord) {
        answerArray.push(" ");
      }
    }
    j--;
  }

  return answerArray.join("");
}
